mailer/templates/templates.go

```diff
@@ -1,124 +1,133 @@
-	TemplateName string `json:"template"`
-	FromName     string `json:"fromName"`
-	Subject      string `json:"subject"`
-	Template     *template.Template
+type Templater struct {
+	RepositoryGitURL string
+	RepositoryBranch string
+	RepositoryFsPath string
+	RepositoryPullInterval time.Duration
+	TheTemplates []markdowntemplates.Template
 }
 
-func (td *LocalizedTemplate) UnmarshalJSON(jsonBytes []byte) error {
-	var t map[string]string
+var theTemplater *Templater
 
-	if err := json.Unmarshal(jsonBytes, &t); err != nil {
-		return err
-	}
-	if err := initStatikFS(); err != nil {
-		return err
-	}
+func NewTemplater(RepositoryGitURL, RepositoryBranch, RepositoryFsPath string, RepositoryPullInterval time.Duration) error {
 
-	f, err := statikFS.Open(fmt.Sprintf("/%s_%s.html", t["template"], t["language"]))
-	if err != nil {
-		return err
-	}
-	templateBytes, err := ioutil.ReadAll(f)
-	if err != nil {
-		return err
+	templater := Templater{
+		RepositoryGitURL: RepositoryGitURL,
+		RepositoryBranch: RepositoryBranch,
+		RepositoryFsPath: RepositoryFsPath,
+		RepositoryPullInterval: RepositoryPullInterval,
 	}
 
-	template, err := template.New(t["template"]).Parse(string(templateBytes))
+	theTemplater = &templater
+
+	err := templater.CloneTemplatesFromGitHub()
 	if err != nil {
 		return err
 	}
-	template.Option("missingkey=error")
-
-	*td = LocalizedTemplate{
-		t["language"],
-		t["template"],
-		t["fromName"],
-		t["subject"],
-		template,
-	}
 
+	// Start routine to fetch latest templates
+	go templater.FetchLatestTemplatesFromGithub()
 	return nil
 }
 
-func initStatikFS() error {
-	sFS, err := fs.New()
-	if err != nil {
-		return err
+func (templater *Templater) FetchLatestTemplatesFromGithub() {
+	ticker := time.NewTicker(templater.RepositoryPullInterval)
+
+	for range ticker.C {
+		cmd := exec.Command("git", "pull", "origin", templater.RepositoryBranch)
+		cmd.Dir = templater.RepositoryFsPath
+		cmd.Stderr = os.Stderr
+		fmt.Printf("Executing %v\n", cmd.Args)
+		err := cmd.Run()
+		if err != nil {
+			fmt.Printf("Error pulling git repository %v\n", err)
+		}
+		err = templater.slurpTemplates(templater.RepositoryFsPath)
+		if err != nil {
+			fmt.Printf("Error reading templates %v\n", err)
+		}
 	}
-	statikFS = sFS
-	return nil
 }
 
-// FromJSON initializes the templates from the templates/templates.json file
-func FromJSON() (int, error) {
-	if err := initStatikFS(); err != nil {
-		return 0, err
+func (templater *Templater) CloneTemplatesFromGitHub() error {
+	// Check if templates folder exists
+	if _, err := os.Stat(templater.RepositoryFsPath); os.IsNotExist(err) {
+		fmt.Println("Templates do not exists; Go and clone repository")
+		cmd := exec.Command("git", "clone", "-b", templater.RepositoryBranch, templater.RepositoryGitURL, templater.RepositoryFsPath)
+		cmd.Stderr = os.Stderr
+		fmt.Printf("Executing %v\n", cmd.Args)
+		return cmd.Run()
 	}
+	return templater.slurpTemplates(templater.RepositoryFsPath)
```

Comment:
I just realized it's working, but only after 5 min when it runs `git pull ...` the first time. So in `CloneTemplatesFromGitHub`, is it executing `templater.slurpTemplates(templater.RepositoryFsPath)` even though it does `return cmd.Run()` beforehand?

Reply:
I think this should be correct, the if block checks if the directory does exist, it clones the repository.

But the observation is really good. Could reproduce that on my machine! I'll investigate later

Reply:
https://github.com/sensebox/sensebox-mailer/pull/14/commits/ea11cf19747e71d172155746743ca619d6e5fc66 could fix that